plugins/core/profile.go

```diff
@@ -0,0 +1,343 @@
+	for _, t := range m.TraceProfileTasks {
+		endTime := t.StartTime + int64(t.Duration)*60*1000
+		if t.EndpointName == endpoint && t.StartTime <= time.Now().UnixMilli() && endTime > time.Now().UnixMilli() && t.Status == reporter.Pending {
+			result = append(result, t)
+		}
+	}
+	return result
+}
+
+func (m *ProfileManager) IfProfiling() bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.status
+}
+
+func (m *ProfileManager) generateProfileLabels(traceSegmentID string, minDurationThreshold int64) profileLabels {
+	var l = &LabelSet{}
+	if minDurationThreshold == 0 {
+		l = Labels(l, SegmentLabel, traceSegmentID)
+	} else {
+		l = Labels(l, SegmentLabel, traceSegmentID, MinDurationLabel, strconv.FormatInt(minDurationThreshold, 10))
+	}
+	closeChan := make(chan struct{}, 1)
+	return profileLabels{
+		labels:    l,
+		closeChan: closeChan,
+	}
+}
+
+func (m *ProfileManager) generateCurrentTask(t *reporter.TraceProfileTask, traceSegmentID string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	var c = currentTask{
+		serialNumber:         t.SerialNumber,
+		taskId:               t.TaskId,
+		traceSegmentId:       traceSegmentID,
+		minDurationThreshold: t.MinDurationThreshold,
+		duration:             t.Duration,
+	}
+	m.currentTask = &c
+}
+
+func (m *ProfileManager) ToProfile(endpoint string, traceSegmentID string) {
+	//check if profiling
+	if m.IfProfiling() {
+		c := m.generateProfileLabels(traceSegmentID, 0)
+		m.labelSets[traceSegmentID] = c
+		SetGoroutineLabels(c.labels)
+		return
+	}
+
+	tasks := m.getProfileTask(endpoint)
+	if tasks != nil {
+		for _, v := range tasks {
+			m.TraceProfileTasks[v.TaskId].Status = reporter.Running
+			//choose task to profiling
+			task := v
+			m.generateCurrentTask(task, traceSegmentID)
+			err := m.StartProfiling(traceSegmentID)
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+			go func(task *reporter.TraceProfileTask) {
+				err = m.monitor()
+				if err != nil {
+					m.TraceProfileTasks[task.TaskId].Status = reporter.Pending
+					m.currentTask = nil
+					m.status = false
+					return
+				}
+				m.TraceProfileTasks[task.TaskId].Status = reporter.Finished
+			}(task)
+			break
+		}
+	}
+
+}
+
+func (m *ProfileManager) StartProfiling(traceSegmentID string) error {
+	m.mu.Lock()
+	m.status = true
+	m.profilingWriter = NewProfilingWriter(
+		ChunkSize,
+		m.rawCh,
+	)
+	// Add main profiling context
+	c := m.generateProfileLabels(traceSegmentID, m.currentTask.minDurationThreshold)
+	SetGoroutineLabels(c.labels)
+	m.labelSets[traceSegmentID] = c
+	m.mu.Unlock()
+
+	if err := pprof.StartCPUProfile(m.profilingWriter); err != nil {
+		m.mu.Lock()
+		m.status = false
+		m.mu.Unlock()
+		return err
+	}
+	return nil
+}
+
+func (m *ProfileManager) monitor() error {
+	select {
+	// End on timeout
+	case <-time.After(time.Duration(m.currentTask.duration) * time.Minute):
+
+	// End manually
+	case <-m.labelSets[m.currentTask.traceSegmentId].closeChan:
+	}
+	// Stop profiling
+	pprof.StopCPUProfile()
```

Comment:
Same here, if there are two requests received, you stopped the CPU profiling when the first endpoint executed, and the second request is not finished. Then, the second request trace profiling is not completed.